File: api/BeSureApi/Models/PartCategory.cs
using System.ComponentModel.DataAnnotations;

namespace BeSureApi.Models
{
    public class PartCategory
    {
        public int Id { get; set; }
        public string Code { get; set; }
        public string Name { get; set; }
        public string ProductCategory { get; set; }
        public int PartProductCategoryId { get; set; }
        public int MappingId { get; set; }
        public string CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public string? UpdatedBy { get; set; }
        public DateTime? UpdatedOn { get; set; }
    }
    public class PartCategoryCreate
    {
        [Required(ErrorMessage = "validation_error_part_category_create_category_name_required")]
        [StringLength(64, ErrorMessage = "validation_error_part_category_create_name_max")]
        public string Name { get; set; }
        [Required(ErrorMessage = "validation_error_part_category_create_product_category_required")]
        public string ProductCategoryId { get; set; }
    }

    public class PartCategoryEdit
    {
        public int Id { get; set; }
        [Required(ErrorMessage = "validation_error_part_category_edit_category_name_required")]
        [StringLength(64, ErrorMessage = "validation_error_part_category_edit_name_max")]
        public string Name { get; set; }
        [Required(ErrorMessage = "validation_error_part_category_edit_product_category_required")]
        public int PartProductCategoryId { get; set; }
        public int MappingId { get; set; }
    }
}

File: ui/src/types/partCategory.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";

export interface PartCategoryCreate {
    Name: string;
    ProductCategoryId: number | string;
}

export interface PartCategoryEdit {
    Id:number;
    Name: string;
    MappingId:number;
    PartProductCategoryId: number | string;
}

export interface PartCategoryCreateResult {
    IsPartCategoryCreated: Boolean;
}

export const partCategoryCreateResultDecoder: Decoder<PartCategoryCreateResult> = object({
    IsPartCategoryCreated: boolean,
});

export interface PartCategoryUpdateResult {
    IsPartCategoryUpdated: Boolean;
}

export const partCategoryUpdateResultDecoder: Decoder<PartCategoryUpdateResult> = object({
    IsPartCategoryUpdated: boolean,
});

export interface PartCategoryDeleted {
    IsPartCategoryUpdated: Boolean;
}

export const partCategoryDeletedDecoder: Decoder<PartCategoryDeleted> = object({
    IsPartCategoryUpdated: boolean,
});

export interface PartCategoryDetails {
    Id: number;
    Code: string;
    Name: string;
    PartProductCategoryId:number;
    ProductCategory: string;
    CreatedBy: string;
    CreatedOn: string;
    UpdatedBy: null | string;
    UpdatedOn: null | string;
    MappingId:number;
}

export interface PartCategoryList {
    PartCategories: PartCategoryDetails[]
    TotalRows: number
    PerPage: number
}

export const partCategoryDetailsDecoder: Decoder<PartCategoryDetails> = object({
    Id: number,
    Code: string,
    Name: string,
    PartProductCategoryId:number,
    ProductCategory: string,
    CreatedBy: string,
    CreatedOn: string,
    UpdatedBy: nullable(string),
    UpdatedOn: nullable(string),
    MappingId:number
})

export const partCategoryListDecoder: Decoder<PartCategoryList> = object({
    PartCategories: array(partCategoryDetailsDecoder),
    TotalRows: number,
    PerPage: number
});

export interface PartCategoryNames {
    Id: number;
    Name: string;
    PartProductCategoryToPartCategoryMappingId: number | null
}

export const partCategoryNamesDecoder: Decoder<PartCategoryNames> = object({
    Id: number,
    Name: string,
    PartProductCategoryToPartCategoryMappingId: nullable(number)
})

export interface PartCategoryNameList {
    PartProductCategoryDetails: PartCategoryNames[]
}


export const partCategoryNameListDecoder: Decoder<PartCategoryNameList> = object({
    PartProductCategoryDetails: array(partCategoryNamesDecoder),
});

export interface PartCategoryName {
    Id: number;
    Code:string;
    Name: string;
}

export const PartCategoryNameDecoder: Decoder<PartCategoryName> = object({
    Id: number,
    Code: string,
    Name: string
});

export interface PartCategoryNameListForPO {
    PartCategoryNames: PartCategoryName[];
}

export const PartCategoryNameListForPODecoder: Decoder<PartCategoryNameListForPO> = object({
    PartCategoryNames: array(PartCategoryNameDecoder),
})

export interface AssetProductCategoryPartCategoryNames {
    Id: number;
    Name: string;
    PartProductCategoryToPartCategoryMappingId: number | null
}

export const assetProductCategoryPartCategoryNamesDecoder: Decoder<AssetProductCategoryPartCategoryNames> = object({
    Id: number,
    Name: string,
    PartProductCategoryToPartCategoryMappingId: nullable(number)
})

export interface AssetProductCategoryPartCategoryNameList {
    PartProductCategoryDetails: AssetProductCategoryPartCategoryNames[]
}


export const AssetProductCategoryPartCategoryNameListDecoder: Decoder<AssetProductCategoryPartCategoryNameList> = object({
    PartProductCategoryDetails: array(assetProductCategoryPartCategoryNamesDecoder),
});